promql/engine_test.go

```diff
@@ -637,27 +637,71 @@ load 10s
 			PeakSamples:  3,
 			Result: Result{
 				nil,
-				Vector{
-					Sample{
-						Point:  Point{V: 0.1, T: 10000},
-						Metric: labels.Labels{},
-					},
-				},
+				Vector{Sample{
+					Point:  Point{V: 0.1, T: 10000},
+					Metric: labels.Labels{},
+				}},
 				nil,
 			},
 			Start: time.Unix(10, 0),
 		},
+		// Multilevel subquery tests for sum(count_over_time(sum_over_time(metric[10s:1s])[10s:]))[10s:2s]
 		{
-			Query:        "sum_over_time(metric[10s:1s])",
-			MaxSamples:   23,
-			TotalSamples: 12,
-			PeakSamples:  23,
+			Query:        "metric[10s:1s]",
+			MaxSamples:   25,
+			TotalSamples: 11,
+			PeakSamples:  11,
+			Result: Result{
+				nil,
+				Matrix{Series{
+					Metric: labels.Labels{labels.Label{Name: "__name__", Value: "metric"}},
+					Points: []Point{{T: 0, V: 1}, {T: 1000, V: 1}, {T: 2000, V: 1}, {T: 3000, V: 1}, {T: 4000, V: 1}, {T: 5000, V: 1}, {T: 6000, V: 1}, {T: 7000, V: 1}, {T: 8000, V: 1}, {T: 9000, V: 1}, {T: 10000, V: 2}},
+				}},
+				nil,
+			},
+			Start: time.Unix(10, 0),
+		},
+		{
+			Query:        "sum_over_time(metric[10s:1s])[10s:]",
+			MaxSamples:   25,
+			TotalSamples: 2,
```

Comment:
This is wrong, it can't be less than the TotalSamples for metric[10s:1s]

Reply:
I think thats because in case of vector selectors, we only calculate during the output. But, here just after evaluating the VS, we go to subquery and then prom function, and hence the samples are selectively chosen based on timestamps, thus reducing the total samples. But, since peak is carried over, so it remains good.

Reply:
So, should we consider adding the samples from `len(res.(Matrix))` (pseudo code) to `totalSamples` after evaluating a subquery?

Reply:
Yes, and ensure than any TotalSamples from that query is also included.